// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package bap;

import "buf/validate/validate.proto";
import "benchmarking/proto/common/metric.proto";
import "benchmarking/proto/common/workflow_type.proto";
import "benchmarking/proto/common/workload_action.proto";

// Defines how frequently the benchmark should be reviewed and/or updated 
// for retention purposes.
enum UpdateFrequency {
  UPDATE_FREQUENCY_UNSPECIFIED = 0;
  WEEKLY = 1;
  MONTHLY = 2;
  QUARTERLY = 3;
}

// Defines the specific execution environment (runner, container image) and stable identity for a benchmark variant.
message EnvironmentConfig {
  // REQUIRED: A unique, stable identifier for this environment configuration (e.g. "gpu_l4_standard").
  string id = 1 [(buf.validate.field).string.min_len = 1];

  // REQUIRED: The GitHub Actions runner label to request (e.g. "linux-x86-n2-32").
  string runner_label = 2 [(buf.validate.field).string.min_len = 1];

  // REQUIRED: The container image to execute the workload inside.
  string container_image = 3 [(buf.validate.field).string.min_len = 1];

  // REQUIRED: The workflow types that this environment should run in (e.g. PRESUBMIT).
  repeated bap.common.WorkflowType workflow_type = 4 [(buf.validate.field).repeated.min_items = 1];
  
  // Environment-specific overrides for workload inputs.
  // These values merge with or override the base workload inputs.
  map<string, string> workload_action_inputs = 5;
}
// Defines a specific benchmark configuration.
message BenchmarkConfig {
  // REQUIRED: A unique name for the benchmark (e.g., "resnet50_jax_train").
  string name = 1 [(buf.validate.field).string.min_len = 1];

  // REQUIRED: A brief description of the benchmark's purpose.
  string description = 2 [(buf.validate.field).string.min_len = 1];
 
  // REQUIRED: The owner or team responsible for the benchmark.
  string owner = 3 [(buf.validate.field).string.min_len = 1];

  // REQUIRED: Configuration for the workload executor action.
  bap.common.WorkloadAction workload = 4 [(buf.validate.field).required = true];

  // REQUIRED: How often this benchmark definition should be reviewed.
  UpdateFrequency update_frequency_policy = 6 [(buf.validate.field).enum = { not_in: [0]}];

  // Optional: GitHub labels for filtering or triggering workflows.
  repeated string github_labels = 7;

  // Optional: List of metrics to track.
  repeated bap.common.MetricSpec metrics = 8;

  // REQUIRED: List of environment configurations.
  repeated EnvironmentConfig environment_configs = 9 [(buf.validate.field).repeated.min_items = 1];

  // Prevent reuse of deleted tags.
  reserved 5;
}

// Represents a collection of benchmark configurations.
message BenchmarkSuite {
  repeated BenchmarkConfig benchmarks = 1;
}
